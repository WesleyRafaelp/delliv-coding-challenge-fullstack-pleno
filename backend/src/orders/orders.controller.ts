import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  UseGuards,
} from '@nestjs/common';
import { OrdersService } from './orders.service';
import { CreateOrderDto } from './dto/create-order.dto';
import { UpdateOrderDto } from './dto/update-order.dto';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';
import { JwtAuthGuard } from 'src/auth/jwt-auth.guard';

@ApiTags('orders')
@Controller('orders')
export class OrdersController {
  constructor(private readonly ordersService: OrdersService) {}

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Post()
  async create(@Body() createOrderDto: CreateOrderDto) {
    return await this.ordersService.create(createOrderDto);
  }

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Get()
  findAll() {
    return this.ordersService.findAll();
  }

  // @Get(':id')
  // findOne(@Param('id') id: string) {
  //   return this.ordersService.findOne(+id);
  // }

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Patch(':id')
  update(@Param('id') id: string, @Body() updateOrderDto: UpdateOrderDto) {
    return this.ordersService.update(+id, updateOrderDto);
  }

  @ApiBearerAuth('JWT-auth')
  @UseGuards(JwtAuthGuard)
  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.ordersService.remove(+id);
  }
}
