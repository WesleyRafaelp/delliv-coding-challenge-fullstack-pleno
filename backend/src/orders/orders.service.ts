import { Injectable } from '@nestjs/common';
import { CreateOrderDto } from './dto/create-order.dto';
import { UpdateOrderDto } from './dto/update-order.dto';
import { PrismaService } from '../prisma.service';

@Injectable()
export class OrdersService {
  constructor(private prisma: PrismaService) {}

  async create(data: CreateOrderDto) {
    const { client, deliveryAddress } = data;

    return await this.prisma.order.create({
      data: {
        client,
        deliveryAddress,
        status: 'PENDENTE',
      },
    });
  }

  findAll() {
    return this.prisma.order.findMany();
  }

  update(id: number, updateOrderDto: UpdateOrderDto) {
    return this.prisma.order.update({
      where: { id },
      data: { status: updateOrderDto.status },
    });
  }

  remove(id: number) {
    return this.prisma.order.delete({ where: { id } });
  }
}
