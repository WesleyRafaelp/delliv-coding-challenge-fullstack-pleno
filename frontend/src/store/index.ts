import { configureStore } from "@reduxjs/toolkit";
import orderReducer  from "./reducers/orders";

const store = configureStore({
    reducer:{
        orderReducer
    }
})

export default store;
export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;