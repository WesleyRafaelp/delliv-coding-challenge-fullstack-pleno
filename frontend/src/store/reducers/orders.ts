import { orderService } from "@/services/order/order-service";
import { Action, Dispatch, ThunkAction, createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { RootState } from "..";
import { stat } from "fs";

export interface Order {
    id: number;
    client: string;
    deliveryAddress: string;
    status: string;
}

interface OrderState {
    orders: Order[];
    loading: boolean;
    error: string | null;
}

const initialState: OrderState = {
    orders: [],
    loading: false,
    error: null,
};

export const getOrders = createAsyncThunk('order/getOrders', orderService.getListOrder);

export const newOrder = createAsyncThunk('order/create', async (order: Partial<Order>) => {
    const res = await orderService.create(order)
    return res
})
export const updateOrder = createAsyncThunk('order/update', async (order: Partial<Order>) => {
    const res = await orderService.update(order)
    return res
})

export const removeOrder = createAsyncThunk('order/remove', async (id: any) => {
    const res = await orderService.delete(id)
    return res
})

export const ordersSlice = createSlice({
    name: "orders",
    initialState,
    reducers: {},
    extraReducers: builder => {
        builder
            .addCase(
                getOrders.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: payload,
                        loading: false,
                        error: null,
                    };
                },
            )
            .addCase(
                getOrders.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                getOrders.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
            .addCase(
                newOrder.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: [...state.orders, payload],
                        error: null,
                        loading: false,
                    };
                },
            )
            .addCase(
                newOrder.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                newOrder.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
            .addCase(
                updateOrder.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: state.orders.map(order =>
                            order.id === payload.id ? { ...order, ...payload } : order
                        ),
                        error: null,
                        loading: false,
                    };
                },
            )
            .addCase(
                updateOrder.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                updateOrder.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
            .addCase(
                removeOrder.fulfilled,
                (state, { payload }) => {
                    return {
                        ...state,
                        orders: state.orders.filter(order => order.id !== payload.id),
                        loading: false,
                        error: null,
                    };
                },
            )
            .addCase(
                removeOrder.pending,
                (state, { payload }) => {
                    return {
                        ...state,
                        loading: true,
                        error: null, 
                    };
                },
            )
            .addCase(
                removeOrder.rejected,
                (state, { payload }: any) => {
                    if (payload && payload.error === 'Unauthorized') {
                        return {
                            ...state,
                            orders: [], 
                            loading: false,
                            error: 'Unauthorized',
                        };
                    }
                    return {
                        ...state,
                        loading: false,
                        error: 'Erro desconhecido',
                    };
                },
            )
    }
})

export default ordersSlice.reducer;
export const orderSelector = (state: RootState) => state.orderReducer;