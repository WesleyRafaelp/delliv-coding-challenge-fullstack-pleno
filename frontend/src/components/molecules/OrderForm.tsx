import React, { useState } from 'react';
import Input from '../atoms/Input';
import Button from '../atoms/Button';
import { newOrder, orderSelector } from '@/store/reducers/orders';
import { useDispatch } from 'react-redux';
import { useAppDispatch, useAppSelector } from '@/hooks/hooks';
import { useForm } from 'react-hook-form';
import { CheckBadgeIcon } from '@heroicons/react/24/solid';

const OrderForm: React.FC = () => {
  const {
    register,
    reset,
    handleSubmit,
    control,
    formState: { errors },
  } = useForm({
    criteriaMode: "all",
  });
  const [isFormSubmitted, setIsFormSubmitted] = useState(false);
  const dispatch = useAppDispatch();
  const { loading, error } = useAppSelector(orderSelector)

  const onSubmit = (data:any) => {

    const dataOrder = {
      client: data.client,
      deliveryAddress: `${data.street}, ${data.number} - ${data.neighborhood}`,
    }
    try {
      dispatch(newOrder(dataOrder));
      setIsFormSubmitted(true);
      reset();
      setTimeout(() => {
        setIsFormSubmitted(false);
      }, 5000); 
    } catch (err) {
      console.error(err)
    }
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className="flex flex-col gap-4 bg-white border border-gray-300 rounded p-4 mb-4">
      <label className="block font-semibold text-gray-700">
        Cliente:
        <Input type="text" placeholder='Digite o nome do cliente' register={register('client', {required: 'Nome do cliente é obrigatório',
        minLength: {
          value: 3,
          message: 'O nome do cliente deve ter pelo menos 3 caracteres.'
        },})} />
      </label>
      {errors.client && (
            <p className="text-red-500">{`${errors.client.message}`}</p>
          )}
      <label className="block font-semibold text-gray-700">
        Rua:        <Input type="text" placeholder='Digite o nome da rua' register={register('street', {required: 'Nome da rua é obrigatório',
        })} />
      </label>
      {errors.street && (
            <p className="text-red-500">{`${errors.street.message}`}</p>
          )}
      <label className="block font-semibold text-gray-700">
        Número :       <Input type="text" placeholder='Digite o número do local' inputMode='numeric'  register={register('number', {required: 'Número residencial é obrigatório', pattern: {
          value: /^\d+$/,
          message: 'Digite apenas números no campo de número.'
        },})} />
      </label>
      {errors.number && (
            <p className="text-red-500">{`${errors.number.message}`}</p>
          )}
      <label className="block font-semibold text-gray-700">
        Bairro :       <Input type="text" placeholder='Digite o nome do bairro' register={register('neighborhood', {required: 'Bairro é obrigatório'})} />
      </label>
      {errors.neighborhood && (
            <p className="text-red-500">{`${errors.neighborhood.message}`}</p>
          )}
      {
        loading ? (<button type="button" className="w-10 h-10" disabled >
          <svg className="animate-spin h-8 w-8 m-auto border-t-2 border-gray rounded-full" viewBox="0 0 24 24"></svg>
        </button>) : (<Button type='submit'>Criar Pedido</Button>)
      }
      {error ? (
        <div className="bg-red-100 border-t-4 border-red-500 rounded-b text-red-900 px-4 py-3 shadow-md mb-4" role="alert" data-testid="error-message">
          <div className="flex">
            <div className="py-1">
              <svg className="fill-current h-6 w-6 text-red-500 mr-4" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20">
                <path
                  d="M10 2C5.82 2 2 5.82 2 10s3.82 8 8 8 8-3.82 8-8-3.82-8-8-8zm1 12h-2v-2h2v2zm0-4h-2V6h2v4z"
                />
              </svg>
            </div>
            <div>
              <p className="font-bold" >Error: {error}</p>
            </div>
          </div>
        </div>
      ): null}
      {isFormSubmitted ? (
        <div className='flex w-full justify-center items-center mt-2'>
          <p className="text-blue-700 font-bold text-center mb-4 flex m-auto">
          Pedido criado com sucesso!
          <CheckBadgeIcon className='w-4 text-blue-700'/>
          </p>
        </div>
      ) : null}
    </form>
  );
};

export default OrderForm;
