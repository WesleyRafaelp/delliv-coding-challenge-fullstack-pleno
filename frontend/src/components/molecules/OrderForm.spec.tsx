import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store  from '@/store/index';
import OrderForm from './OrderForm';
import { newOrder } from '@/store/reducers/orders';
import { useAppDispatch } from '@/hooks/hooks';
import { useAppSelector } from '@/hooks/hooks';

jest.mock('../../store/reducers/orders', () => ({
  __esModule: true,
  newOrder: jest.fn(),
}));

jest.mock('../../hooks/hooks', () => ({
  ...jest.requireActual('../../hooks/hooks'),
  useAppDispatch: jest.fn(),
  useAppSelector: jest.fn(),
}));

describe('OrderForm Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders OrderForm component', () => {
    const mockDispatch = jest.fn();
    useAppDispatch.mockReturnValue(mockDispatch);
    useAppSelector.mockReturnValue(mockDispatch);
    render(
      <Provider store={store}>
        <OrderForm />
      </Provider>
    );
    expect(screen.getByLabelText(/cliente/i)).toBeInTheDocument();
  });

  it('submits the form and dispatches newOrder action', async () => {
    const mockDispatch = jest.fn();
    useAppDispatch.mockReturnValue(mockDispatch);
    useAppSelector.mockReturnValue(mockDispatch);

    render(
      <Provider store={store}>
        <OrderForm />
      </Provider>
    );

    fireEvent.change(screen.getByLabelText(/cliente/i), { target: { value: 'John Doe' } });
    fireEvent.change(screen.getByLabelText(/rua/i), { target: { value: 'Main Street' } });
    fireEvent.change(screen.getByLabelText(/número/i), { target: { value: '123' } });
    fireEvent.change(screen.getByLabelText(/bairro/i), { target: { value: 'Downtown' } });

    fireEvent.click(screen.getByText(/criar pedido/i));

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalled();
      expect(newOrder).toHaveBeenCalledWith({
        client: 'John Doe',
        deliveryAddress: 'Main Street, 123 - Downtown',
      });
    });
  });

  it('displays error messages for empty fields', async () => {
    render(
      <Provider store={store}>
        <OrderForm />
      </Provider>
    );

    fireEvent.click(screen.getByText(/criar pedido/i));

    await waitFor(() => {
      expect(screen.getByText(/nome do cliente é obrigatório/i)).toBeInTheDocument();
      expect(screen.getByText(/nome da rua é obrigatório/i)).toBeInTheDocument();
      expect(screen.getByText(/número residencial é obrigatório/i)).toBeInTheDocument();
      expect(screen.getByText(/bairro é obrigatório/i)).toBeInTheDocument();
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <OrderForm />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays success message after form submission', async () => {
    render(
      <Provider store={store}>
        <OrderForm />
      </Provider>
    );

    fireEvent.change(screen.getByLabelText(/cliente/i), { target: { value: 'John Doe' } });
    fireEvent.change(screen.getByLabelText(/rua/i), { target: { value: 'Main Street' } });
    fireEvent.change(screen.getByLabelText(/número/i), { target: { value: '123' } });
    fireEvent.change(screen.getByLabelText(/bairro/i), { target: { value: 'Downtown' } });

    fireEvent.click(screen.getByText(/criar pedido/i));

    await waitFor(() => {
      expect(screen.getByText(/pedido criado com sucesso/i)).toBeInTheDocument();
    });
  });
  it('displays error message if newOrder action fails', async () => {
    const mockDispatch = jest.fn(() => {
      throw new Error('Some error');
    });
    useAppSelector.mockReturnValue(mockDispatch);
    useAppDispatch.mockReturnValue(mockDispatch);

    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
  
    render(
      <Provider store={store}>
        <OrderForm />
      </Provider>
    );
  
    fireEvent.change(screen.getByLabelText(/cliente/i), { target: { value: 'John Doe' } });
    fireEvent.change(screen.getByLabelText(/rua/i), { target: { value: 'Main Street' } });
    fireEvent.change(screen.getByLabelText(/número/i), { target: { value: '123' } });
    fireEvent.change(screen.getByLabelText(/bairro/i), { target: { value: 'Downtown' } });
  
    fireEvent.click(screen.getByText(/criar pedido/i));
  
    await waitFor(() => {
     expect(consoleErrorSpy).toHaveBeenCalledWith(expect.any(Error));
    expect(consoleErrorSpy.mock.calls[0][0].message).toContain('Some error');
    });
  
    
    consoleErrorSpy.mockRestore();
  });

});
