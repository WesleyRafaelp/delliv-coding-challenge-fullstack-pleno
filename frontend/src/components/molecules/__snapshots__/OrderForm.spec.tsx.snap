// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-4 bg-white border border-gray-300 rounded p-4 mb-4"
  >
    <label
      class="block font-semibold text-gray-700"
    >
      Cliente:
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        name="client"
        placeholder="Digite o nome do cliente"
        type="text"
        value=""
      />
    </label>
    <label
      class="block font-semibold text-gray-700"
    >
      Rua:        
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        name="street"
        placeholder="Digite o nome da rua"
        type="text"
        value=""
      />
    </label>
    <label
      class="block font-semibold text-gray-700"
    >
      Número :       
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        inputmode="numeric"
        name="number"
        placeholder="Digite o número do local"
        type="text"
        value=""
      />
    </label>
    <label
      class="block font-semibold text-gray-700"
    >
      Bairro :       
      <input
        class="border border-gray-300 rounded p-2 w-full focus:outline-none focus:ring focus:border-blue-300"
        name="neighborhood"
        placeholder="Digite o nome do bairro"
        type="text"
        value=""
      />
    </label>
    <button
      class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline-blue active:bg-blue-800"
      type="submit"
    >
      Criar Pedido
    </button>
  </form>
</DocumentFragment>
`;
